core/src/main/java/com/linecorp/armeria/client/retry/HttpStatusBasedRetryStrategy.java

```diff
@@ -44,13 +46,18 @@
     }
 
     @Override
-    public CompletionStage<Backoff> shouldRetry(HttpRequest request, HttpResponse response) {
-        final CompletableFuture<HttpHeaders> future = new CompletableFuture<>();
-        final HttpHeaderSubscriber subscriber = new HttpHeaderSubscriber(future);
-        response.completionFuture().whenComplete(subscriber);
-        response.subscribe(subscriber);
-
-        return future.handle((headers, thrown) -> backoffFunction
-                .apply(headers != null ? headers.status() : null, thrown));
+    public CompletionStage<Backoff> shouldRetry(ClientRequestContext ctx, @Nullable HttpResponse response,
+                                                @Nullable Throwable cause) {
+        if (cause instanceof UnprocessedRequestException) {
+            // We can just retry immediately when the UNprocessedRequestException is raised.
+            return CompletableFuture.completedFuture(Backoff.withoutDelay());
```

Comment:
There's a chance of thrashing a server in this case. A server may be in graceful-shutdown phase, rejecting any incoming requests with GOAWAY, which will make the server suffer from a high spike of retries. How about consulting `backoffFunction`?

Reply:
Oops my big mistake. Fixed. Thank you~!